File: autojs/src/main/java/com/stardust/autojs/engine/RhinoJavaScriptEngine.java
package com.stardust.autojs.engine;

import android.util.Log;
import android.view.View;

import com.stardust.autojs.BuildConfig;
import com.stardust.autojs.core.ui.ViewExtras;
import com.stardust.autojs.rhino.NativeJavaObjectWithPrototype;
import com.stardust.autojs.rhino.RhinoAndroidHelper;
import com.stardust.autojs.rhino.TopLevelScope;
import com.stardust.autojs.runtime.ScriptRuntime;
import com.stardust.autojs.script.JavaScriptSource;
import com.stardust.autojs.script.StringScriptSource;
import com.stardust.automator.UiObjectCollection;
import com.stardust.pio.PFiles;
import com.stardust.pio.UncheckedIOException;

import org.mozilla.javascript.Context;
import org.mozilla.javascript.Script;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;
import org.mozilla.javascript.commonjs.module.RequireBuilder;
import org.mozilla.javascript.commonjs.module.provider.SoftCachingModuleScriptProvider;

import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.Collections;
import java.util.Locale;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by Stardust on 2017/4/2.
 */

public class RhinoJavaScriptEngine extends JavaScriptEngine {

    public static final String SOURCE_NAME_INIT = "<init>";

    private static final String LOG_TAG = "RhinoJavaScriptEngine";

    private static final String MODULES_PATH = "modules";
    private static Script sInitScript;
    private static final ConcurrentHashMap<Context, RhinoJavaScriptEngine> sContextEngineMap = new ConcurrentHashMap<>();

    private Context mContext;
    private TopLevelScope mScriptable;
    private Thread mThread;
    private android.content.Context mAndroidContext;

    public RhinoJavaScriptEngine(android.content.Context context) {
        mAndroidContext = context;
        mContext = enterContext();
        mScriptable = createScope(mContext);
    }

    @Override
    public void put(String name, Object value) {
        ScriptableObject.putProperty(mScriptable, name, Context.javaToJS(value, mScriptable));
    }

    @Override
    public void setRuntime(ScriptRuntime runtime) {
        super.setRuntime(runtime);
        runtime.setTopLevelScope(mScriptable);
    }

    @Override
    public Object doExecution(JavaScriptSource source) {
        Reader reader = source.getNonNullScriptReader();
        try {
            reader = preprocess(reader);
            Script script = mContext.compileReader(reader, source.toString(), 1, null);
            return mContext.executeScriptWithContinuations(script, mScriptable);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    protected Reader preprocess(Reader script) throws IOException {
        return script;
    }

    @Override
    public void forceStop() {
        Log.d(LOG_TAG, "forceStop: interrupt Thread: " + mThread);
        mThread.interrupt();
    }


    @Override
    public synchronized void destroy() {
        super.destroy();
        Log.d(LOG_TAG, "on destroy");
        sContextEngineMap.remove(getContext());
        Context.exit();
    }

    public Thread getThread() {
        return mThread;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void init() {
        mThread = Thread.currentThread();
        ScriptableObject.putProperty(mScriptable, "__engine__", this);
        initRequireBuilder(mContext, mScriptable);
        mContext.executeScriptWithContinuations(getInitScript(), mScriptable);
    }

    private Script getInitScript() {
        if (sInitScript == null || BuildConfig.DEBUG) {
            try {
                Reader reader = new InputStreamReader(mAndroidContext.getAssets().open("init.js"));
                sInitScript = mContext.compileReader(reader, SOURCE_NAME_INIT, 1, null);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
        return sInitScript;
    }

    void initRequireBuilder(Context context, Scriptable scope) {
        AssetAndUrlModuleSourceProvider provider = new AssetAndUrlModuleSourceProvider(mAndroidContext, MODULES_PATH,
                Collections.singletonList(new File("/").toURI()));
        new RequireBuilder()
                .setModuleScriptProvider(new SoftCachingModuleScriptProvider(provider))
                .setSandboxed(true)
                .createRequire(context, scope)
                .install(scope);

    }

    public Context getContext() {
        return mContext;
    }

    public Scriptable getScriptable() {
        return mScriptable;
    }

    protected TopLevelScope createScope(Context context) {
        TopLevelScope topLevelScope = new TopLevelScope();
        topLevelScope.initStandardObjects(context, false);
        return topLevelScope;
    }

    public Context enterContext() {
        Context context = new RhinoAndroidHelper(mAndroidContext).enterContext();
        setupContext(context);
        sContextEngineMap.put(context, this);
        return context;
    }

    protected void setupContext(Context context) {
        context.setOptimizationLevel(-1);
        context.setLanguageVersion(Context.VERSION_ES6);
        context.setLocale(Locale.getDefault());
        context.setWrapFactory(new WrapFactory());
    }


    public static RhinoJavaScriptEngine getEngineOfContext(Context context) {
        return sContextEngineMap.get(context);
    }

    private class WrapFactory extends org.mozilla.javascript.WrapFactory {

        @Override
        public Object wrap(Context cx, Scriptable scope, Object obj, Class<?> staticType) {
            Object result;
            if (obj instanceof String) {
                result = getRuntime().bridges.toString(obj.toString());
            } else if (staticType == UiObjectCollection.class) {
                result = getRuntime().bridges.asArray(obj);
            } else {
                result = super.wrap(cx, scope, obj, staticType);
            }
            return result;
        }

        @Override
        public Scriptable wrapAsJavaObject(Context cx, Scriptable scope, Object javaObject, Class<?> staticType) {
            Scriptable result;
            if (javaObject instanceof View) {
                result = ViewExtras.getNativeView(scope, (View) javaObject, staticType, getRuntime());
            } else {
                result = super.wrapAsJavaObject(cx, scope, javaObject, staticType);
            }
            //Log.d(LOG_TAG, "wrapAsJavaObject: java = " + javaObject + ", result = " + result + ", scope = " + scope);
            return result;
        }

    }


}

File: autojs/src/main/java/com/stardust/autojs/engine/RhinoJavaScriptEngine.kt
package com.stardust.autojs.engine

import android.util.Log
import android.view.View
import com.stardust.autojs.core.ui.ViewExtras
import com.stardust.autojs.engine.module.AssetAndUrlModuleSourceProvider
import com.stardust.autojs.execution.ExecutionConfig
import com.stardust.autojs.project.ScriptConfig
import com.stardust.autojs.rhino.RhinoAndroidHelper
import com.stardust.autojs.rhino.TopLevelScope
import com.stardust.autojs.runtime.ScriptRuntime
import com.stardust.autojs.runtime.exception.ScriptInterruptedException
import com.stardust.autojs.script.JavaScriptSource
import com.stardust.automator.UiObjectCollection
import com.stardust.pio.UncheckedIOException
import org.mozilla.javascript.Context
import org.mozilla.javascript.Script
import org.mozilla.javascript.Scriptable
import org.mozilla.javascript.ScriptableObject
import org.mozilla.javascript.commonjs.module.RequireBuilder
import org.mozilla.javascript.commonjs.module.provider.SoftCachingModuleScriptProvider
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader
import java.lang.ref.WeakReference
import java.net.URI
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by Stardust on 2017/4/2.
 */

open class RhinoJavaScriptEngine(private val mAndroidContext: android.content.Context) :
    JavaScriptEngine() {

    private val context: WeakReference<Context>
    private val mScriptable: WeakReference<TopLevelScope>
    private var unsealed: Boolean = false
    lateinit var threadRef: WeakReference<Thread>

    private val initScript: Script
        get() {
            return sInitScript ?: {
                try {
                    val reader = InputStreamReader(mAndroidContext.assets.open("init.js"))
                    val script = context.get()?.compileReader(reader, SOURCE_NAME_INIT, 1, null)
                    sInitScript = script
                    script
                } catch (e: IOException) {
                    throw UncheckedIOException(e)
                }
            }()!!
        }

    fun getContext(): Context? {
        return context.get()
    }

    fun getThread(): Thread? {
        return threadRef.get()
    }

    val scriptable: Scriptable
        get() = mScriptable.get()!!

    init {
        this.context = WeakReference(enterContext())
        mScriptable = this.context.get()?.let { createScope(it) }!!
        this.context.get()?.seal(mScriptable)
    }

    override fun put(name: String, value: Any?) {
        ScriptableObject.putProperty(
            mScriptable.get(),
            name,
            Context.javaToJS(value, mScriptable.get())
        )
    }

    override fun setRuntime(runtime: ScriptRuntime) {
        super.setRuntime(runtime)
        runtime.topLevelScope = mScriptable.get()
    }

    public override fun doExecution(source: JavaScriptSource): Any? {
        var reader = source.nonNullScriptReader
        try {
            reader = preprocess(reader)
            val script = context.get()?.compileReader(reader, source.toString(), 1, null)
            return if (hasFeature(ScriptConfig.FEATURE_CONTINUATION)) {
                context.get()?.executeScriptWithContinuations(script, mScriptable.get())
            } else {
                script?.exec(context.get(), mScriptable.get())
            }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }

    }

    fun hasFeature(feature: String): Boolean {
        val config = getTag(ExecutionConfig.tag) as ExecutionConfig?
        return config != null && config.scriptConfig.hasFeature(feature)
    }

    @Throws(IOException::class)
    protected fun preprocess(script: Reader): Reader {
        return script
    }

    override fun forceStop() {
        Log.d(LOG_TAG, "forceStop: interrupt Thread: ${threadRef.get()}")
        threadRef.get()?.interrupt()
    }


    @Synchronized
    override fun destroy() {
        var destroySuccess = false;
        try {
            super.destroy()
            Log.d(LOG_TAG, "on destroy")
            destroySuccess = true;
        } finally {
            if (destroySuccess)
                Log.d(LOG_TAG, "destroy execute success")
            else
                Log.d(LOG_TAG, "destroy execute failed")
            sContextEngineMap.remove(context.get())
            Context.exit()
        }
    }

    override fun init() {
        threadRef = WeakReference(Thread.currentThread())
        ScriptableObject.putProperty(mScriptable.get(), "__engine__", this)
        mScriptable.get()?.let { context.get()?.let { ctx -> initRequireBuilder(ctx, it) } }
        try {
            context.get()?.executeScriptWithContinuations(initScript, mScriptable.get())
        } catch (e: IllegalArgumentException) {
            if ("Script argument was not a script or was not created by interpreted mode " == e.message) {
                initScript.exec(context.get(), mScriptable.get())
            } else {
                throw e
            }
        }
    }

    internal fun initRequireBuilder(context: Context, scope: Scriptable) {
        val provider = AssetAndUrlModuleSourceProvider(
            mAndroidContext, MODULES_PATH,
            listOf<URI>(File("/").toURI())
        )
        RequireBuilder()
            .setModuleScriptProvider(SoftCachingModuleScriptProvider(provider))
            .setSandboxed(true)
            .createRequire(context, scope)
            .install(scope)

    }

    protected fun createScope(context: Context): WeakReference<TopLevelScope> {
        val topLevelScope = TopLevelScope()
        topLevelScope.initStandardObjects(context, false)
        if (unsealed) {
            // FIXME 会导致无法正常被回收 存在内存泄露可能
            topLevelScope.setNoRecycle()
        }
        return WeakReference(topLevelScope)
    }

    fun enterContext(): Context {
        val context = RhinoAndroidHelper(mAndroidContext).enterContext()
        setupContext(context)
        sContextEngineMap[context] = this
        return context
    }

    fun exitContext(context: Context) {
        sContextEngineMap.remove(context)
        try {
            Context.exit()
        } catch (e: Exception) {
            // do nothing
        }
    }

    protected fun setupContext(context: Context) {
        // FIXME 同时跑多个UI脚本时会共用同一个Context 此时的WrapFactory会被覆盖
        if (context.isSealed) {
            unsealed = true
            context.unseal(sContextEngineMap[context]?.mScriptable)
        } else {
            context.optimizationLevel = -1
            context.languageVersion = Context.VERSION_ES6
            context.locale = Locale.getDefault()
            context.wrapFactory = WrapFactory()
        }
    }

    private inner class WrapFactory : org.mozilla.javascript.WrapFactory() {

        override fun wrap(cx: Context, scope: Scriptable, obj: Any?, staticType: Class<*>?): Any? {
            return when {
                obj is String -> runtime.bridges.toString(obj.toString())
                staticType == UiObjectCollection::class.java -> runtime.bridges.asArray(obj)
                else -> {
                    if (scope is TopLevelScope) {
                        if (scope.isRecycled) {
                            throw ScriptInterruptedException()
                        }
                    }
                    super.wrap(cx, scope, obj, staticType)
                }
            }
        }

        override fun wrapAsJavaObject(
            cx: Context?,
            scope: Scriptable,
            javaObject: Any?,
            staticType: Class<*>?
        ): Scriptable? {
            //Log.d(LOG_TAG, "wrapAsJavaObject: java = " + javaObject + ", result = " + result + ", scope = " + scope);
            return if (javaObject is View) {
                ViewExtras.getNativeView(scope, javaObject, staticType, runtime)
            } else {
                super.wrapAsJavaObject(cx, scope, javaObject, staticType)
            }
        }

    }

    companion object {

        val SOURCE_NAME_INIT = "<init>"

        private val LOG_TAG = "RhinoJavaScriptEngine"

        private val MODULES_PATH = "modules"
        private var sInitScript: Script? = null
        private val sContextEngineMap = ConcurrentHashMap<Context, RhinoJavaScriptEngine>()


        fun getEngineOfContext(context: Context): RhinoJavaScriptEngine? {
            return sContextEngineMap[context]
        }
    }


}
